### PR TITLE
Header validation

### PR DESCRIPTION
**Pull request by samay-kothari**
Added the functions that would be required for header validation at the time of header download. The changes to existing functions are commented until these functions are incorporated completely so that the functioning is not affected.

### CODE REVIEW

**Comment by kcalvinalvin** [comment]
> The changes to existing functions are commented until these functions are incorporated completely so that the functioning is not affected.

There shouldn't be any changes to the existing functions since that would make the node inoperable. The CI failing shows that.

There would also need to be a unit test for the newly added function.

Also, the copyright headers should be on the top of each file.

**Comment by samay-kothari** [comment]
I have removed all the additional comments to the existing functions, also removed the implementation of bestHeader from the blockIndex file, still the CI test is failing, can you check the file changes once and guide me. Also for the unit tests, what will be the procedure to add them?

File: blockchain/checkpoints.go [hunk 1]
```diff
@@ -259,3 +260,31 @@ func (b *BlockChain) IsCheckpointCandidate(block *btcutil.Block) (bool, error) {
 	// All of the checks passed, so the block is a candidate.
 	return true, nil
 }
+
+// maybeUpdateMostRecentCheckpoint potentially updates the most recently known
+// checkpoint to the provided block node.
+//
+// This function MUST be called with the chain lock held (for writes).
+func (b *BlockChain) maybeUpdateMostRecentCheckpoint(node *blockNode) {
```
Comment (kcalvinalvin) [inline blockchain/checkpoints.go:268]: This introduces a hard fork. We don't want this function for our code.

File: blockchain/blockindex.go [hunk 2]
```diff
@@ -228,6 +306,7 @@ type blockIndex struct {
 	// separate mutex.
 	db          database.DB
 	chainParams *chaincfg.Params
+	bestHeader  *blockNode
```
Comment (kcalvinalvin) [inline blockchain/blockindex.go:309]: Avoid adding bestHeader for now since this isn't needed for `ProcessBlockHeader()`. This and all associated code should be removed in this PR.

File: blockchain/process.go [hunk 3]
```diff
@@ -129,6 +131,137 @@ func (b *BlockChain) processOrphans(hash *chainhash.Hash, flags BehaviorFlags) e
...
+// returned.
+//
+// It should be noted that this function intentionally does not accept block
+// headers that do not connect to an existing known header or to headers which
+// are already known to be a part of an invalid branch.  This means headers must
+// be processed in order.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) ProcessBlockHeader(header *wire.BlockHeader) error {
+	b.chainLock.Lock()
```
Comment (kcalvinalvin) [inline blockchain/process.go:253]: Prefer `defer b.chainLock.Unlock()`. The decred code had to do call `b.chainLock.Unlock()` since defers are stacked and it was already calling a defer with `processLock` which isn't present in this repo.

File: blockchain/process.go [hunk 4]
```diff
@@ -129,6 +131,137 @@ func (b *BlockChain) processOrphans(hash *chainhash.Hash, flags BehaviorFlags) e
...
+	// Note that the additional information for the actual votes, tickets, and
+	// revocations in the block can't be populated until the full block data is
+	// known since that information is not available in the header.
+	newNode := newBlockNode(header, prevNode)
+	newNode.status = statusNone
+	b.index.AddNode(newNode)
+
+	// Potentially update the most recently known checkpoint to this block
+	// header.
+	b.maybeUpdateMostRecentCheckpoint(newNode)
```
Comment (kcalvinalvin) [inline blockchain/process.go:231]: As I've mentioned in the function definition, this shouldn't be here since it causes a hard fork.

File: blockchain/process.go [hunk 5]
```diff
@@ -129,6 +131,137 @@ func (b *BlockChain) processOrphans(hash *chainhash.Hash, flags BehaviorFlags) e
...
+// accept its header in order to quickly eliminate blocks that are obviously
+// incorrect.
+//
+// In the case the block header is already known, the associated block node is
+// examined to determine if the block is already known to be invalid, in which
+// case an appropriate error will be returned.  Otherwise, the block node is
+// returned.
+//
+// This function MUST be called with the chain lock held (for writes).
+func (b *BlockChain) maybeAcceptBlockHeader(header *wire.BlockHeader, checkHeaderSanity bool) (*blockNode, error) {
```
Comment (kcalvinalvin) [inline blockchain/process.go:174]: I'd prefer putting this function in `accept.go`

**Review by kcalvinalvin — CHANGES_REQUESTED** [review]
The tests [AcceptHeader()](https://github.com/decred/dcrd/blob/f0efcc2226fe15814911390af42f35e72d540a6a/internal/blockchain/common_test.go#L655-L659) and [RejectHeader()](https://github.com/decred/dcrd/blob/f0efcc2226fe15814911390af42f35e72d540a6a/internal/blockchain/common_test.go#L778-L781) should also be included in this PR along with a modified [TestProcessLogic()](https://github.com/decred/dcrd/blob/f0efcc2226fe15814911390af42f35e72d540a6a/internal/blockchain/process_test.go#L521) (that only includes what we need in this PR).

File: blockchain/checkpoints.go (hunk 1 above)
Comment (samay-kothari) [reply blockchain/checkpoints.go]: Removed

File: blockchain/process.go (hunk 3 above)
Comment (samay-kothari) [reply blockchain/process.go:171]: Made the changes

File: blockchain/process.go (hunk 4 above)
Comment (samay-kothari) [reply blockchain/process.go]: Removed the function

File: blockchain/process.go (hunk 5 above)
Comment (samay-kothari) [reply blockchain/process.go]: changed the location of the functions

**Comment by kcalvinalvin** [comment]
Could you rebase on main?

**Comment by kcalvinalvin** [comment]
Could you remove the commits `bfb96eff6e07d95a06c0267c877b79d51187d4bb` and `32ee9e65cac1797de04f34177f17c7559968cec2`. They are not needed and it makes reviewing more tedious as I have to sort through those files.

You should also squash commits that are just removing code from the commit from before like commit `bf98308a71c2b3936f332e21e113a13dc55c2bbc`. How to [squash](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits)

**Comment by samay-kothari** [comment]
Hi @kcalvinalvin 
I removed and squashed the redundant commits, and reduced them to three, kindly check.

File: blockchain/blockindex.go [hunk 6]
```diff
@@ -303,13 +302,6 @@ func (bi *blockIndex) AddNode(node *blockNode) {
 // This function is NOT safe for concurrent access.
 func (bi *blockIndex) addNode(node *blockNode) {
 	bi.index[node.hash] = node
-	if bi.bestHeader == nil {
```
Comment (kcalvinalvin) [inline blockchain/blockindex.go]: These removes should be in the previous commit.

File: blockchain/checkpoints.go [hunk 7]
```diff
@@ -260,31 +260,3 @@ func (b *BlockChain) IsCheckpointCandidate(block *btcutil.Block) (bool, error) {
 	// All of the checks passed, so the block is a candidate.
 	return true, nil
 }
-
```
Comment (kcalvinalvin) [inline blockchain/checkpoints.go]: Same goes for this code block. This should be squashed into the previous commit

**Comment by kcalvinalvin** [comment]
General note: every git commit should ideally be "atomic", meaning it should be addressing one feature/bug and not multiple.

In commit `8bf4075de664ddb12be3e4ff9248c620a81ba7ec`  you've added unnecessary and you remove them in commit `91a5ea851d6dd881cb33bbb2e3ea4eb7a4b302aa`. This makes it harder to navigate through the commits. `8bf4075de664ddb12be3e4ff9248c620a81ba7ec` itself doesn't build either so that should be addressed so that `8bf4075de664ddb12be3e4ff9248c620a81ba7ec` builds without `91a5ea851d6dd881cb33bbb2e3ea4eb7a4b302aa`

**Comment by samay-kothari** [comment]
Hi @kcalvinalvin 
This makes sense
I have tried to sort out the commits order and made them according the functioning, really sorry for the issues.

**Comment by kcalvinalvin** [comment]
> Hi @kcalvinalvin
> This makes sense
> I have tried to sort out the commits order and made them according the functioning, really sorry for the issues.

The commits I asked to be changed are not there though?

**Comment by samay-kothari** [comment]
> The commits I asked to be changed are still there though?

I have removed them now. Cleared the history

**Comment by samay-kothari** [comment]
hi @kcalvinalvin I have made the changes that you asked for in the call, please check.

File: blockchain/process.go [hunk 8]
```diff
@@ -129,6 +131,54 @@ func (b *BlockChain) processOrphans(hash *chainhash.Hash, flags BehaviorFlags) e
...
+// This function is safe for concurrent access.
+func (b *BlockChain) ProcessBlockHeader(header *wire.BlockHeader) error {
+	b.chainLock.Lock()
+	defer b.chainLock.Unlock()
+	const checkHeaderSanity = true
+	_, err := b.maybeAcceptBlockHeader(header, checkHeaderSanity)
+	if err != nil {
+		return err
+	}
+
```
Comment (kcalvinalvin) [inline blockchain/process.go:178]: Since `maybeAcceptBlockHeader()` made changes to the blockindex, we should be flushing the index to disk with the below function.
https://github.com/utreexo/utreexod/blob/7e21d3d7ba843a7559e771091c373b88dc7dc07b/blockchain/blockindex.go#L324

File: blockchain/chain.go [hunk 9]
```diff
@@ -1364,6 +1364,20 @@ func (b *BlockChain) BlockLocatorFromHash(hash *chainhash.Hash) BlockLocator {
 	return locator
 }
 
+func (b *BlockChain) IndexLookupNode(hash *chainhash.Hash) *blockNode {
```
Comment (kcalvinalvin) [inline blockchain/chain.go:1367]: An exported function should always have a comment for documentation. https://go.dev/doc/comment

File: blockchain/chain.go [hunk 10]
```diff
@@ -1364,6 +1364,20 @@ func (b *BlockChain) BlockLocatorFromHash(hash *chainhash.Hash) BlockLocator {
...
+	b.chainLock.RLock()
+	node := b.index.LookupNode(hash)
+	b.chainLock.RUnlock()
+	return node
+}
+
+func (b *BlockChain) IndexNodeStatus(node *blockNode) blockStatus {
```
Comment (kcalvinalvin) [inline blockchain/chain.go:1374]: An exported function should always have a comment for documentation. https://go.dev/doc/comment

File: blockchain/chain.go [hunk 11]
```diff
@@ -1389,6 +1403,21 @@ func (b *BlockChain) BlockHeightByHash(hash *chainhash.Hash) (int32, error) {
 	return node.height, nil
 }
 
+// NodeHeightByHash returns the height of the block with the given hash in the
+// main chain. Does not check if the node is on the main chain, so returns height
+// of the node which is created by header validation
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) NodeHeightByHash(hash *chainhash.Hash) (int32, error) {
```
Comment (kcalvinalvin) [inline blockchain/chain.go:1411]: Why do we need  this function? This is only called in test code and you can do `node := LookupNode` anyways.

File: blockchain/fullblocktests/generate.go [hunk 12]
```diff
@@ -175,10 +175,10 @@ func makeSpendableOut(block *wire.MsgBlock, txIndex, txOutIndex uint32) spendabl
 	return makeSpendableOutForTx(block.Transactions[txIndex], txOutIndex)
 }
 
-// testGenerator houses state used to easy the process of generating test blocks
+// TestGenerator houses state used to easy the process of generating test blocks
```
Comment (kcalvinalvin) [inline blockchain/fullblocktests/generate.go:178]: Exporting of the TestGenerator should be in a separate commit.

File: blockchain/common_test.go [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+)
+
+type chaingenHarness struct {
+	*fullblocktests.TestGenerator
+
+	t     *testing.T
+	chain *blockchain.BlockChain
+}
+
+func (g *chaingenHarness) AcceptHeader(blockName string) {
```
Comment (kcalvinalvin) [inline blockchain/common_test.go:17]: Documentation should exist about AcceptHeader here

File: blockchain/process.go (hunk 8 above)
Comment (samay-kothari) [reply blockchain/process.go:178]: yeah this makes sense, but this should be done in maybeAcceptBlockHeader() function right?
Just as we are doing it in maybeAcceptBlock() function here: https://github.com/utreexo/utreexod/blob/7e21d3d7ba843a7559e771091c373b88dc7dc07b/blockchain/accept.go#L70-L74

File: blockchain/chain.go (hunk 11 above)
Comment (samay-kothari) [reply blockchain/chain.go:1411]: This right now is being used while testing only, but it is also used to get height of the headers, when we are executing sliding block window logic.

File: blockchain/fullblocktests/generate.go (hunk 12 above)
Comment (samay-kothari) [reply blockchain/fullblocktests/generate.go:178]: Done

File: blockchain/chain.go (hunk 10 above)
Comment (samay-kothari) [reply blockchain/chain.go:1374]: Okay, will add

**Comment by samay-kothari** [comment]
Hi calvin,
I have made the requested changes in the PR, please check

File: blockchain/chain.go (hunk 10 above)
Comment (kcalvinalvin) [reply blockchain/chain.go:1381]: The comment is wrong. The function is `IndexNodeStatus` but the comment is saying `NodeStatus`.

File: blockchain/common_test.go [hunk 14]
```diff
@@ -0,0 +1,129 @@
+package blockchain_test
```
Comment (kcalvinalvin) [inline blockchain/common_test.go:1]: What was your reasoning for naming this package `blockchain_test`?

File: blockchain/process_test.go [hunk 15]
```diff
@@ -0,0 +1,46 @@
+package blockchain_test
```
Comment (kcalvinalvin) [inline blockchain/process_test.go:1]: Same for this file. What was your reasoning for naming this `blockchain_test`?

File: blockchain/chain.go (hunk 10 above)
Comment (samay-kothari) [reply blockchain/chain.go:1381]: Ohh, I am really sorry, fixed that

File: blockchain/common_test.go (hunk 14 above)
Comment (samay-kothari) [reply blockchain/common_test.go:1]: keeping this as part of blockchain package causes import cycle
So I took reference from fullblocks_test and example_test files which are part of the package blockchain_test
https://github.com/utreexo/utreexod/blob/7e21d3d7ba843a7559e771091c373b88dc7dc07b/blockchain/fullblocks_test.go#L6

File: blockchain/process_test.go (hunk 15 above)
Comment (samay-kothari) [reply blockchain/process_test.go:1]: Same reasoning as it is for file common_test.go

File: blockchain/common_test.go (hunk 14 above)
Comment (kcalvinalvin) [reply blockchain/common_test.go:1]: 2 things:

1. In Go, the package name must match the directory the file is in. So for a package `blockchain_test`, you'd have to make a new directory.
2. I'd like to avoid new packages and I don't think the tests being added here really need a new package. If import cycles are a problem, you could the tests into package fullblocktests. You'd have to bring back the HeightByHash function for this.

Comment (samay-kothari) [reply blockchain/common_test.go:1]: If we take the second approach, then we won't need to export testGenerator as well, so we can remove that commit
But we need to export ChainSetup(change chainSetup to ChainSetup) from blockchain package.

Comment (samay-kothari) [reply blockchain/common_test.go:1]: Also should I put the fullblocks_test and example_test files which exists already as part of blockchain_test package in fullblockstest package?

Comment (kcalvinalvin) [reply blockchain/common_test.go:1]: > If we take the second approach, then we won't need to export testGenerator as well, so we can remove that commit
But we need to export ChainSetup(change chainSetup to ChainSetup) from blockchain package.

Sounds good. Should do the exporting in a separate commit as well.

> Also should I put the fullblocks_test and example_test files which exists already as part of blockchain_test package in fullblockstest package?

Hmm ok I did some searching and apparently having a `_test` package is used at times. Let's keep the existing tests where they are for now and just move `AcceptHeader` and `RejectHeader` tests into fullblocktests package since these rely on the TestGenerator a lot.

**Comment by samay-kothari** [comment]
I don't know why it failed the test here, on my forked repository, this commit passed the tests
https://github.com/samay-kothari/utreexod/actions/runs/2703305958
Maybe some error came up, re-run might help

File: blockchain/accept.go [hunk 16]
```diff
@@ -90,3 +92,84 @@ func (b *BlockChain) maybeAcceptBlock(block *btcutil.Block, flags BehaviorFlags)
 
 	return isMainChain, nil
 }
+
+// maybeAcceptBlockHeader potentially accepts the header to the block index and,
+// if accepted, returns the block node associated with the header.  It performs
+// several context independent checks as well as those which depend on its
+// position within the chain.  It should be noted that some of the header fields
```
Comment (kcalvinalvin) [inline blockchain/accept.go:99]: Which header fields require the full block data to be able to validate? This looks like a exact copy paste of the comments from dcrd.

Comment (samay-kothari) [reply blockchain/accept.go:99]: Yeah I will remove that part.
The header fields that they are talking about are some data about voters and stakes, which we are not maintaining here

File: blockchain/accept.go [hunk 17]
```diff
@@ -90,3 +92,84 @@ func (b *BlockChain) maybeAcceptBlock(block *btcutil.Block, flags BehaviorFlags)
...
+		return nil, err
+	}
+
+	// Create a new block node for the block and add it to the block index.
+	//
+	// Note that the additional information for the actual votes, tickets, and
+	// revocations in the block can't be populated until the full block data is
+	// known since that information is not available in the header.
+	newNode := newBlockNode(header, prevNode)
+	newNode.status = statusNone
```
Comment (kcalvinalvin) [inline blockchain/accept.go:168]: This line doesn't need to be here since we're setting new blockNodes to be status none now

File: blockchain/accept.go [hunk 18]
```diff
@@ -90,3 +92,84 @@ func (b *BlockChain) maybeAcceptBlock(block *btcutil.Block, flags BehaviorFlags)
...
+	b.index.AddNode(newNode)
+	err = b.index.flushToDB()
```
Comment (kcalvinalvin) [inline blockchain/accept.go:170]: So looking into this more, I'd prefer having the flush happening in `ProcessBlockHeader` since we'd eventually want to be able to re-use this function in `ProcessBlock` and `ProcessBlock` takes care of the flush itself.

File: blockchain/accept.go (hunk 17 above)
Comment (samay-kothari) [reply blockchain/accept.go:168]: Yeah makes sense, will remove it

**Comment by kcalvinalvin** [comment]
The commit messages should ideally follow the golang [contribution guide](https://go.dev/doc/contribute#commit_messages).

For example, commit `9933f5e7ffb6b5f7b0b15825602493f77ee13baa` should be something like:

```
blockchain: add for header validation function

ProcessBlockHeader allows for separately validating and accepting
block headers which is needed for headers-first block downloads.
```

File: blockchain/accept.go (hunk 18 above)
Comment (samay-kothari) [reply blockchain/accept.go:170]: okay that makes sense, since we would be using maybeAcceptBlockHeader in ProcessBlock when we would receive a block whose header haven't been validated.

Comment (samay-kothari) [reply blockchain/accept.go:170]: But do we need header validation, if we have the complete block data to us and we are using maybeAcceptBlock function to validate block anyways.

Comment (samay-kothari) [reply blockchain/accept.go:170]: Anyways I will add the dbFlush to ProcessBlockHeader function

**Comment by samay-kothari** [comment]
Hi calvin,
I have made the requested changes and also edited the commit messages, according to the standards, kindly check.

File: blockchain/fullblocktests/generate.go [hunk 19]
```diff
@@ -756,6 +756,16 @@ func (g *testGenerator) assertTipBlockMerkleRoot(expected chainhash.Hash) {
 	}
 }
 
+// BlockByName returns the block associated with the provided block name.  It
```
Comment (kcalvinalvin) [inline blockchain/fullblocktests/generate.go:759]: `BlockByName` should be lowercase since the function isn't exported

Comment (samay-kothari) [reply blockchain/fullblocktests/generate.go]: Okay, made the change, check please

File: blockchain/fullblocktests/generate.go [hunk 20]
```diff
@@ -906,6 +916,10 @@ func Generate(includeLargeReorg bool) (tests [][]TestInstance, err error) {
 		})
 	}
 
+	// Checking that duplicate blocks are not allowed, done
```
Comment (kcalvinalvin) [inline blockchain/fullblocktests/generate.go:919]: What changes were made to the `HaveBlock` function?

Comment (samay-kothari) [reply blockchain/fullblocktests/generate.go:919]: Changes in HaveBlock would be required when we would be integrating it into the working code.
No need in this PR

File: blockchain/fullblocktests/generate.go [hunk 21]
```diff
@@ -2133,3 +2147,142 @@ func Generate(includeLargeReorg bool) (tests [][]TestInstance, err error) {
 
 	return tests, nil
 }
+
+type TestHeaderInstance interface {
```
Comment (kcalvinalvin) [inline blockchain/fullblocktests/generate.go:2151]: There's no comments on any of these exported structs and functions

Comment (samay-kothari) [reply blockchain/fullblocktests/generate.go]: Added comments and moved the header structs along with the block structs, so all of them can be referred. And added comments for the functions inside GenerateHeaders

File: blockchain/fullblocktests/generate.go (hunk 20 above)
Comment (kcalvinalvin) [reply blockchain/fullblocktests/generate.go:957]: Ok then shouldn't it be removed?

Comment (samay-kothari) [reply blockchain/fullblocktests/generate.go]: I have added this test here from a different PR, will edit the comment.
Right now this test just checks for duplicate block, which was not being done earlier in fullblockstest

File: blockchain/fullblocktests/common_test.go [hunk 22]
```diff
@@ -0,0 +1,128 @@
+package fullblocktests
```
Comment (kcalvinalvin) [inline blockchain/fullblocktests/common_test.go:1]: Missing header license info.

Should add

```
// Copyright (c) 2022 The utreexod developers
// Use of this source code is governed by an ISC
// license that can be found in the LICENSE file.
```

File: blockchain/fullblocktests/process_test.go [hunk 23]
```diff
@@ -0,0 +1,46 @@
+package fullblocktests
```
Comment (kcalvinalvin) [inline blockchain/fullblocktests/process_test.go:1]: Also missing license header here

File: blockchain/fullblocktests/generate.go [hunk 24]
```diff
@@ -2133,3 +2183,134 @@ func Generate(includeLargeReorg bool) (tests [][]TestInstance, err error) {
 
 	return tests, nil
 }
+
+func GenerateHeaders() (generator *testGenerator, tests [][]TestInstance) {
```
Comment (kcalvinalvin) [inline blockchain/fullblocktests/generate.go:2187]: Could you leave a comment here about how `GenerateHeaders` works? It's not intuitive because the function itself is not really generating headers, rather it's generating blocks that'll only be used for headers.

Comment (kcalvinalvin) [inline blockchain/fullblocktests/generate.go:2187]: We've added two new errors in this PR: `ErrKnownInvalidBlock` and `ErrorMissingParent`. Could you add reject tests for those cases as well?

Comment (samay-kothari) [reply blockchain/fullblocktests/generate.go:2187]: To test these error, we need integrated functionality of header validation and block validation, which is not present right now, I have added tests for these errors in next PR. 
But they are necessary since we are checking for these errors while validating headers.

Comment (samay-kothari) [reply blockchain/fullblocktests/generate.go:2187]: Okay I will add that

File: blockchain/fullblocktests/generate.go [hunk 25]
```diff
@@ -756,6 +794,16 @@ func (g *testGenerator) assertTipBlockMerkleRoot(expected chainhash.Hash) {
 	}
 }
 
+// blockByName returns the block associated with the provided block name.  It
+// will panic if the specified block name does not exist.
+func (g *testGenerator) blockByName(blockName string) *wire.MsgBlock {
```
[the remark this answers is not in the document]
Comment (samay-kothari) [reply blockchain/fullblocktests/generate.go:799]: I have used this function, only in the code that we have introduced in this PR. For existing code, wherever we are using blocksByName[blockName], those functions are part of testGenerator struct only, so I thought, there is no need to change there.
I will add the function at these two places, which are part of the Generate function
https://github.com/utreexo/utreexod/blob/7e21d3d7ba843a7559e771091c373b88dc7dc07b/blockchain/fullblocktests/generate.go#L890

https://github.com/utreexo/utreexod/blob/7e21d3d7ba843a7559e771091c373b88dc7dc07b/blockchain/fullblocktests/generate.go#L2110

File: blockchain/fullblocktests/generate.go [hunk 26]
```diff
@@ -98,6 +98,44 @@ var _ TestInstance = RejectedBlock{}
 // This implements the TestInstance interface.
 func (b RejectedBlock) FullBlockTestInstance() {}
 
+// AcceptedHeader defines a test instance that expects a header to be accepted to
+// the blockchain either by extending the main chain, on a side chain, but can't be
+// an orphan
+type AcceptedHeader struct {
+	Name        string
+	BlockHeader *wire.MsgBlock
```
[the remark this answers is not in the document]
Comment (samay-kothari) [reply blockchain/fullblocktests/generate.go:106]: Yeah basically BlockHeader represent blocks with just header, so I named it as Block header, I will just change the name to Block

File: blockchain/fullblocktests/generate.go [hunk 27]
```diff
@@ -98,6 +98,44 @@ var _ TestInstance = RejectedBlock{}
...
+	Height      int32
+	IsMainChain bool
```
[the remark this answers is not in the document]
Comment (samay-kothari) [reply blockchain/fullblocktests/generate.go:108]: I thought we would these need parameters, when we will add block data to these headers, but we will have to create different struct for that, we can remove IsMainChain and isOrphan from AcceptHeader struct.

[the remark this answers is not in the document]
Comment (kcalvinalvin) [reply blockchain/fullblocktests/generate.go]: But the blocks downloaded will match the headers, and the headers are only downloaded from a pre-hardcoded set that's included in the binary. The main advantage with headers-first is that we don't download orphan blocks and that we stay away from lower work chains when we first sync.

[the remark this answers is not in the document]
Comment (samay-kothari) [reply blockchain/fullblocktests/generate.go]: Yes, it makes sense, I have removed these two parameters as we won't be using them in future as well.

File: blockchain/fullblocktests/generate.go (hunk 24 above)
Comment (kcalvinalvin) [reply blockchain/fullblocktests/generate.go:2189]: Ok you're correct in that you can't test for `ErrKnownInvalidBlock` until we have the block validation integration but `ErrorMissingParent` is able to be tested currently.

```diff
diff --git a/blockchain/fullblocktests/generate.go b/blockchain/fullblocktests/generate.go
index 45a1a7ed..fdb08830 100644
--- a/blockchain/fullblocktests/generate.go
+++ b/blockchain/fullblocktests/generate.go
@@ -2264,6 +2264,11 @@ func GenerateHeaders() (generator *testGenerator, tests [][]TestInstance) {
        //                                      \-> b1(1)
        g.setTip("b0")
        b1 := g.nextBlock("b1", outs[1])
+
+       // Create a block with the parent as a header that will be rejected.
+       // This block's header will trigger the the error "missing parent".
+       g.nextBlock("b1a", outs[1])
+
        // This can't be done inside a munge function passed to nextBlock
        // because the block is solved after the function returns and this test
        // requires an unsolved block.
@@ -2281,7 +2286,12 @@ func GenerateHeaders() (generator *testGenerator, tests [][]TestInstance) {
                }
                g.updateBlockState("b46", origHash, "b46", b1)
        }
+       g.setTip("b1")
        rejected(blockchain.ErrHighHash)
+
+       g.setTip("b1a")
+       rejected(blockchain.ErrMissingParent)
+
        // Create block with an invalid proof-of-work limit.
        //
        //   ... -> b0()
```

File: blockchain/fullblocktests/generate.go [hunk 28]
```diff
@@ -2133,3 +2181,138 @@ func Generate(includeLargeReorg bool) (tests [][]TestInstance, err error) {
...
+	// involves an unsolvable block.
+	{
+		origHash := b3.BlockHash()
+		b3.Header.Bits = 0x01810000 // -1 in compact form.
+		g.updateBlockState("b3", origHash, "b3", b3)
+	}
+	rejected(blockchain.ErrUnexpectedDifficulty)
+	// Adding a block with valid header
+	//
+	// ... -> b0() -> b4(2)
```
Comment (kcalvinalvin) [inline blockchain/fullblocktests/generate.go:2313]: Why was b4(1) changed to b4(2)?

Comment (samay-kothari) [reply blockchain/fullblocktests/generate.go:2313]: I changed it earlier, to test immature spend error, but that was not being detected on headers download, so changed it back in the code, forgot to change the comment.

File: blockchain/fullblocktests/generate.go (hunk 24 above)
Comment (samay-kothari) [reply blockchain/fullblocktests/generate.go:2189]: Yeah this test will work, I will add this and update the commit.

**Comment by kcalvinalvin** [comment]
Ok looks good! Good work :)
